Content.Server/_Harmony/JoinQueue/JoinQueueManager.cs

```diff
@@ -0,0 +1,169 @@
+
+        if (!value)
+        {
+            foreach (var session in _queue)
+                session.Channel.Disconnect("Queue was disabled");
+        }
+    }
+
+
+    private async void OnPlayerStatusChanged(object? sender, SessionStatusEventArgs e)
+    {
+        if (e.NewStatus == SessionStatus.Disconnected)
+        {
+            var wasInQueue = _queue.Remove(e.Session);
+
+            if (!wasInQueue && e.OldStatus != SessionStatus.InGame) // Process queue only if player disconnected from InGame or from queue
+                return;
+
+            ProcessQueue(true, e.Session.ConnectedTime);
+
+            if (wasInQueue)
+                QueueTimings.WithLabels("Unwaited").Observe((DateTime.UtcNow - e.Session.ConnectedTime).TotalSeconds);
+        }
+        else if (e.NewStatus == SessionStatus.Connected)
+        {
+            OnPlayerConnected(e.Session);
+        }
+    }
+
+
+    private async void OnPlayerConnected(ICommonSession session)
+    {
+        if (!_isEnabled)
+        {
+            SendToGame(session);
+            return;
+        }
+
+        var isPrivileged = await _connection.HasPrivilegedJoin(session.UserId);
+        var currentOnline = _player.PlayerCount - 1;
+        var haveFreeSlot = currentOnline < _configuration.GetCVar(CCVars.SoftMaxPlayers);
+        if (isPrivileged || haveFreeSlot)
+        {
+            SendToGame(session);
+        }
+        else
+        {
+            _queue.Add(session);
+        }
+
+        ProcessQueue(false, session.ConnectedTime);
+    }
+
+    /// <summary>
+    /// If possible, takes the first player in the queue and sends him into the game
+    /// </summary>
+    /// <param name="isDisconnect">Is method called on disconnect event</param>
+    /// <param name="connectedTime">Session connected time for histogram metrics</param>
+    private void ProcessQueue(bool isDisconnect, DateTime connectedTime)
+    {
+        var players = ActualPlayersCount;
+        if (isDisconnect)
+            players--; // Decrease currently disconnected session but that has not yet been deleted
+
+        var haveFreeSlot = players < _configuration.GetCVar(CCVars.SoftMaxPlayers);
+        var regularQueueContains = _queue.Count > 0;
+
+        if (haveFreeSlot && regularQueueContains)
+        {
+            ICommonSession session = _queue.First();
+            SendToGame(session);
+            QueueTimings.WithLabels("Waited").Observe((DateTime.UtcNow - connectedTime).TotalSeconds);
+        }
+
+        SendUpdateMessages();
+        QueueCount.Set(_queue.Count);
+    }
+
+    /// <summary>
+    /// Sends messages to all players in the queue with the current state of the queue
+    /// </summary>
+    private void SendUpdateMessages()
+    {
+        var totalInQueue = _queue.Count;
+        var currentPosition = 1;
+
+        for (var i = 0; i < _queue.Count; i++, currentPosition++)
+        {
+            _queue[i].Channel.SendMessage(new QueueUpdateMessage
```

Comment:
My editor gives me a warning because a line break is missing for whatever reason
```suggestion
            _queue[i]
                .Channel.SendMessage(new QueueUpdateMessage
```